crates/task-impls/src/helpers.rs

```diff
@@ -37,32 +38,64 @@ use hotshot_types::{
 use tokio::task::JoinHandle;
 use tracing::{debug, error, info, instrument, warn};
 
-use crate::{
-    events::{HotShotEvent, ProposalMissing},
-    request::REQUEST_TIMEOUT,
-};
+use crate::{events::HotShotEvent, request::REQUEST_TIMEOUT};
 
-/// Trigger a request to the network for a proposal for a view and wait for the response
+/// Trigger a request to the network for a proposal for a view and wait for the response or timeout.
 #[instrument(skip_all)]
 pub(crate) async fn fetch_proposal<TYPES: NodeType>(
-    view: TYPES::Time,
-    event_stream: Sender<Arc<HotShotEvent<TYPES>>>,
+    view_number: TYPES::Time,
+    event_sender: Sender<Arc<HotShotEvent<TYPES>>>,
+    event_receiver: Receiver<Arc<HotShotEvent<TYPES>>>,
     quorum_membership: Arc<TYPES::Membership>,
     consensus: OuterConsensus<TYPES>,
+    sender_key: TYPES::SignatureKey,
 ) -> Result<Leaf<TYPES>> {
-    let (tx, mut rx) = broadcast(1);
-    let event = ProposalMissing {
-        view,
-        response_chan: tx,
-    };
+    // First, broadcast that we need a proposal to the current leader
     broadcast_event(
-        Arc::new(HotShotEvent::QuorumProposalRequest(event)),
-        &event_stream,
+        HotShotEvent::QuorumProposalRequestSend(view_number, sender_key).into(),
+        &event_sender,
     )
     .await;
-    let Ok(Ok(Some(proposal))) = async_timeout(REQUEST_TIMEOUT, rx.recv_direct()).await else {
+
+    // Make a background task to await the arrival of the event data.
+    let Ok(Some(proposal)) =
+        // We want to explicitly timeout here so we aren't waiting around for the data.
+        async_timeout(REQUEST_TIMEOUT, async move {
+            // First, capture the output from the event dependency
+            let event = EventDependency::new(
+                event_receiver.clone(),
+                Box::new(move |event| {
+                    let event = event.as_ref();
+                    if let HotShotEvent::QuorumProposalResponseRecv(
+                        qpr_view_number,
+                        _quorum_proposal,
+                    ) = event
+                    {
+                        *qpr_view_number == view_number
+                    } else {
+                        false
+                    }
+                }),
+            )
+            .completed()
+            .await;
```

Comment:
Similar to the comment in `message.rs` - we could get rid of the need for an extra task if we send and forget. But not sure how nicely that plays with the dependency stuff

Reply:
The downside of send-and-forget is that it disallows us the ability to block on the data coming back. So, in this case, the proposal's arrival begets a block before we vote/propose if we're missing that prior state. Because of that, I opted to just block (as the prior version did as well) and timeout in some short interval if we don't hear back from the leader in time. I think in general though, I agree that this type of blocking should *not* happen. I'd like to do a more earnest look at the R/R tasks but I didn't want to make this a massive PR.

Reply:
Gotcha, yeah. Maybe down the line we could utilize the dependency framework to get it to block - although admittedly I am not the most familiar. For this PR it is fine though